### ISSUE TITLE
Union can be compared to any value

### ISSUE THREAD

**Issue by ProTip** [Microsoft/TypeScript#14854]
For some reason I can compare a union to anything without complaint from the compiler. I would expect it to complain that the thing I'm comparing it to is not a valid option.

**Code**

```ts
function do_stuff(stuff: {} | {}[]) {
  if ( stuff === String) {}
}
```

**Expected behavior:**
This should throw an error because we know string constructor is not an option of the union.

**Actual behavior:**
No errors.

**Comment by DanielRosenwasser** [comment]
When using `===`, we check whether either side is assignable to the other. `String` is a proper subtype of the empty object type (`{}`), so this seems appropriate.

**Comment by ProTip** [comment]
For what it's worth I get this with ```==``` as well. Is this expected?

Is there no way to describe the constraint in the current type system? I'm expecting a map or array of maps essentially; unfortunately most libraries simply return indexable objects. I'm not expecting a String constructor.

I could accept a Map if not for the fact that most libraries out there do simply work with objects...

**Comment by RyanCavanaugh** [comment]
>  I'm expecting a map or array of maps essentially; unfortunately most libraries simply return indexable objects. I'm not expecting a String constructor.

But... the `String` constructor *is* a map (`{ }`). The distinction here is one of intent rather than structure. There's not really a bright line for "Things which are key-value lookups" vs "Things which are objects with fixed keys".

You could write this instead, but it may present problems to your callers:
```ts
function do_stuff<T>(stuff: { [s: string]: {} } | {}[]) {
  if ( stuff === String) {}
}
```
